src/main/kotlin/JobState.kt

```diff
@@ -78,17 +87,63 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
         return status
     }
 
-    suspend fun process(newStatus: JobStatus, predicate: suspend (U) -> Boolean, msg: String? = null): JobState<U> {
+    /**
+     * Transforms the type of the job by using the supplied transform function.
+     * If an unhandled exception is encountered, sets the JobState to TransientFailure.
+     */
+    inline fun <R: Any> map(transform: (U) -> R): JobState<R>  = processMap(JobStatus.TransientFailure, transform)
+
+    /**
+     * Transforms the type of the job by using the supplied transform function.
+     * If an unhandled exception is encountered, sets the JobState to PermanentFailure.
+     */
+    inline fun <R: Any> mapRequire(transform: (U) -> R): JobState<R> = processMap(JobStatus.PermanentFailure, transform)
+
+    @PublishedApi
+    internal inline fun <R: Any> processMap(newStatus: JobStatus, transform: (U) -> R): JobState<R>{
+        try{
+            val transFormedVal: R? = when (inProgress()) {
+                true -> transform(value!!)
+                false -> null
+            }
+            return JobState(transFormedVal, this.interceptors).also { it.status = this.status }
+        }catch(e: Exception) {
+            logger.debug { "Failed to mapRequire: ${e.message}" }
+            return JobState<R>(null, interceptors).also { it.status = newStatus }
+        }
+    }
+
+    private suspend fun processWorkerFunction(fn: suspend(U) -> WorkerResult, msg: String? = null): JobState<U> {
+        val lastInterceptor = WorkerInterceptor {
+            if(inProgress()) {
+                val result = fn(value!!)
+                if(FAILED_JOB_STATUSES.contains(result)){
```

Comment:
This could be done by adding a member to the Enum itself,

But that approach has both advantages and disadvantages... So I'm happy with this